### DISCUSSION TITLE
Fancy left prompt recipes, examples, etc

### DISCUSSION THREAD

**Discussion by micimize** [akinomyoga/ble.sh#282] [Q&A] [answered]
Hi – been loving blesh so far.

One thing I've been missing from zsh is powerlevel9k/powerlevel10k-style prompt configs.

After fiddling a little with the airline feature I'm guessing ble.sh might already have some of it's capabilities, so I thought I'd ask if there were any known recipes?

Also wondered if it would be feasible to have prompt elements spec approach sometime in the future:
```sh
typeset -g POWERLEVEL9K_LEFT_PROMPT_ELEMENTS=(
  virtualenv time dir background_jobs_joined vcs vi_mode
)
p10k reload
```
<img width="466" alt="image" src="https://user-images.githubusercontent.com/8343799/220213752-76d3a33e-2165-46f6-a8e4-6ce5f1698e3e.png">

**Comment by akinomyoga** [comment, marked as answer]
Thank you for the suggestion.

> I'm guessing ble.sh might already have some of it's capabilities,

`ble.sh` itself doesn't have the feature. As described in README, `ble.sh` has been intended to be a line editor as a replacement for Readline, which is independent of prompt settings. My intention was that `ble.sh` would be combined with the user's prompt setting or with another framework primarily focusing on prompts. One of the reasons that I prepared vim-airline is for the demonstration of the status bar, and another reason is that there are no other prompt settings for Bash that provides the status bar because the status bar is `ble.sh`'s specific feature.

On the other hand, I would assume that there would be already many prompt settings for `PS1` that are carefully designed. However, I'm actually not familiar with a good one (partly because I personally prefer a very simple prompt like `PS1='[\u@\h \W]\$ '`). The following is a list of the projects that provides an *element/segment*-based prompt setting, but I haven't tested whether they work correctly under `ble.sh`. Also, I'm not sure if they are implemented in an efficient way to minimize the delay.

- agnoster [speedenator/agnoster-bash: Agnoster Theme for Bash](https://github.com/speedenator/agnoster-bash)
- powerline [Bash-it/bash-it: A community Bash framework.](https://github.com/Bash-it/bash-it) with `BASH_IT_THEME=powerline` (or `powerline-plain`, `powerline-multiline`, or `powerline-naked`)
- pureline [chris-marsh/pureline: A Pure Bash Powerline PS1 Command Prompt](https://github.com/chris-marsh/pureline)
- sbp [brujoand/sbp: Simple Bash Prompt (SBP)](https://github.com/brujoand/sbp)
- starship [starship/starship: ☄🌌️ The minimal, blazing-fast, and infinitely customizable prompt for any shell!](https://github.com/starship/starship) ... However, this is not written in Bash. It calls an external program through the command substitution `$(/path/to/starship ...)`.

Of course, I guess there are even more segment/element-based prompt settings that I'm not aware of. And, also I think there are many other powerline-style prompts that are not segment/element-based.

But maybe I can implement a good one, which is efficient, adaptive, and better fused with ble.sh, if I have time.

**Reply by micimize** [reply]
Ah this is perfect! I'll look into these. Thanks!

**Reply by micimize** [reply]
Thought I'd come back and mention that I've started using `starship` and find it very nice. All changes to their config file can be seen the very next line, and the config even has some toml schema thing where it'll auto-complete what you're typing. Only downside is it's in rust, which I love but might be a bit of a hassle if I want to try and beam my setup onto remote servers for work
